streams/src/main/java/org/apache/kafka/streams/processor/internals/assignment/Graph.java

```diff
@@ -0,0 +1,377 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.kafka.streams.processor.internals.assignment;
+
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.Map;
+import java.util.Map.Entry;
+import java.util.Objects;
+import java.util.Set;
+import java.util.SortedMap;
+import java.util.SortedSet;
+import java.util.TreeMap;
+import java.util.TreeSet;
+import java.util.stream.Collectors;
+
+public class Graph<V extends Comparable<V>> {
+    public class Edge {
+        final V destination;
+        final int capacity;
+        final int cost;
+        int residualFlow;
```

Comment:
Do we actually need this field?

If I read the code right, for a forward edge, it's always `capacity - flow` and for a backward edge it's always `0`. So it seem redundant (and potentially error prone to store it expliclity)? -- Instead we could have a `residualFlow()` method that compute it on-the-fly (we could also simplify the update logic when modifying flow as we only need to update the `flow` itself)? Or do I read the update logic inside `cancelNegativeCycle` incorrectly and those properties are not an invariant?

Reply:
> for a backward edge it's always 0

`residualFlow` for backward edge is the flow in forward edge: https://github.com/apache/kafka/blob/trunk/streams/src/main/java/org/apache/kafka/streams/processor/internals/assignment/Graph.java#L178

Reply:
Ah. Parameter order... Thought it's `, flow, residualFlow,` (not sure why you picked the "reverse" order)